clang/test/SemaCXX/deduced-return-type-cxx14.cpp

```diff
@@ -640,3 +640,36 @@ namespace PR46637 {
   template<typename T> struct Y { T x; };
   Y<auto() -> auto> y; // expected-error {{'auto' not allowed in template argument}}
 }
+
+namespace GH71015 {
+
+// Check that there is no error in case a templated function is recursive and
+// has a placeholder return type.
+struct Node {
+  int value;
+  Node* left;
+  Node* right;
+};
+
+bool parse(const char*);
+Node* parsePrimaryExpr();
+
+auto parseMulExpr(auto node) { // cxx14-error {{'auto' not allowed in function prototype}}
+  if (node == nullptr) node = parsePrimaryExpr();
+  if (!parse("*")) return node;
+  return parseMulExpr(new Node{.left = node, .right = parsePrimaryExpr()});
```

Comment:
It would be great to add a test that this is rejected without the previous line `retuen node;`

Reply:
Added the test.

A note here is that It will only be rejected if the function instantiated because clang doesn't take into account return statements if the function is templated due to this part:

https://github.com/llvm/llvm-project/blob/11276563c81987791a2326950dbc3315a32dd709/clang/lib/Sema/SemaStmt.cpp#L3806

gcc rejects even if the function is not instantiated. msvc doesn't. https://compiler-explorer.com/z/hGc8M1esT

Reply:
I think that's fine, and it's what I expected. While we could try to reject in the template definition if we see a recursive call prior to the end of the first `return` statement, I don't think it's worth the added complexity.

I *think* what GCC's doing here is that it's treating the recursive call as dependent in much the same way that Clang does with this patch, but then performing an instantiation of the function template specialization `parseMulExpr<Node*>` *after* it finishes processing the template definition -- so it's also not eagerly diagnosing the problem. (Clang suppresses performing recursive non-dependent function template instantiations until the template has a point of instantiation outside itself, which GCC doesn't seem to do.) But I could be wrong.